### ISSUE TITLE
Bug: [DeepPick] on nullable type

### ISSUE THREAD

**Issue by AndreasHald** [ts-essentials/ts-essentials#380]
### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript`                           | `X.Y.Z` |
| `ts-essentials`                        | `X.Y.Z` |

### Repro Code

```typescript
import { DeepPick } from 'ts-essentials';

interface EnglishClass {
  teacher: {
    name: string;
    yearsOfExperience: number;
  } | null;
}

type EnglishClassNamesOnly = DeepPick<
  // ^?
  EnglishClass,
  {
    teacher: {
      name: never;
    };
  }
>;
```

### TypeScript Playground

https://www.typescriptlang.org/play?ssl=7&ssc=12&pln=8&pc=2#code/JYWwDg9gTgLgBAbzgEQKarABWAYwNZwC+cAZlBCHAOQwDOAtKrbagHYzACGANrVQNwBYAFAjg7VFBKccqOAFFWAc27BaACwDC3Ts0Qi4cGKhnrJALn3DDh1pxCpLtGFHFKh1mwE8TUWgHkSeQAPMElgNllLVgBXEAAjSQ9DYgAfOFjubg9CEREYLzCFZVUNbV1aADl7Jn9Wbi84AF4UdCxcPAAeAzgAel64AD0Afh7FFTUtHWYAGh6EHsNjUwsrGxs7B2jUADckxaJkohEAPg8gA

### Expected Result

I expected a type to be constructed that would be 
```ts 
teacher: {
name: string
} | null
```

### Actual Result

TS error

**Comment by AndreasHald** [comment]
It actually looks like it constructs the type properly - however a TS error is still thrown

![CleanShot 2024-01-20 at 14 05 37@2x](https://github.com/ts-essentials/ts-essentials/assets/8914722/a39341a8-15fe-45d3-a037-285a50f0cc41)

**Comment by Beraliv** [comment]
Hey @AndreasHald

Thank you for reporting the issue, it looks like `DeepPick` doesn't support nullable properties because it shows the generic constraint is `{ teacher?: true | undefined; }` and doesn't contain an option with partial `Teacher` object

I apologise in advance but I won't be available next 2 weeks. When I'm back, I can have a look at this

**Comment by Beraliv** [comment]
Hey again @AndreasHald, looking at this issue right now

The issue that you're facing is related to the changes that were made with `DeepPick` and `DeepOmit` last year. The current problem is that the generic constraint is failing inferring the right structure for the second type argument for you therefore you see TypeScript threw TypeError. The inferred value will be still correct.

The similar bug was reported here but for a generic type - https://github.com/ts-essentials/ts-essentials/issues/343

To unblock, I'd like to introduce some changes (will be a minor release) described in the comment - https://github.com/ts-essentials/ts-essentials/issues/343#issuecomment-2041170333

I will keep you posted about it

**Comment by Beraliv** [comment]
@AndreasHald I've created a PR - https://github.com/ts-essentials/ts-essentials/pull/384, so please let me know if you have any input

**Comment by AndreasHald** [comment]
@Beraliv Thanks for creating a PR - If I understand the solution correctly I would simply stick to `DeepPick` and no ts error would be thrown? and `StrictDeepPick` is then for the other use cases?

I'm afraid the level of Typescript in the library is beyond my capabilities to critique others than "works as I expected"/"doesn't work as I expected".

So if this is the correct solution I believe you :) In my opinion this can for sure be a minor release.

**Comment by Beraliv** [comment]
@AndreasHald it is optimal solution for you as you'd like to see the result, validation is an optional part that I don't want to enforce until there is a safe option available.

One exception to what I said would be "I will introduce it as a major change" as it's technically breaking the contract that existed. I've also discussed it in another issue and agreed to make it major - https://github.com/ts-essentials/ts-essentials/issues/343#issuecomment-2042090016

**Comment by Beraliv** [comment]
The PR - https://github.com/ts-essentials/ts-essentials/pull/384 was just merged, the changes will be released by the end of April as major

**Comment by Beraliv** [comment]
It's released just now as a part of [v10.0.0](https://github.com/ts-essentials/ts-essentials/releases/tag/v10.0.0)

**Comment by AndreasHald** [comment]
@Beraliv fixed my issue :) thanks so much!
